From a3e6680f935a0b6533563c0b58d41a217996aa93 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Thu, 1 Dec 2016 20:53:13 +0000
Subject: [PATCH] [CONFIGURATION-641] Improved exception thrown by load().

XMLConfiguration.load() now checks whether the locator has been
correctly initialized. If not, an exception with a helpful message
is thrown.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1772266 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration2/XMLConfiguration.java      |  7 ++++++
 .../configuration2/TestXMLConfiguration.java  | 22 +++++++++++++++++++
 2 files changed, 29 insertions(+)

diff --git a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
index 72a012ec9b..bc2fc6b8e1 100644
--- a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
@@ -962,6 +962,13 @@ public void read(InputStream in) throws ConfigurationException, IOException
      */
     private void load(InputSource source) throws ConfigurationException
     {
+        if (locator == null)
+        {
+            throw new ConfigurationException("Load operation not properly "
+                    + "initialized! Do not call read(InputStream) directly,"
+                    + " but use a FileHandler to load a configuration.");
+        }
+
         try
         {
             URL sourceURL = locator.getSourceURL();
diff --git a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
index 5ec2d38412..979892e51d 100644
--- a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
@@ -1621,6 +1621,28 @@ public void testPublicIdSynchronized()
                 Methods.END_READ);
     }
 
+    /**
+     * Tests a direct invocation of the read() method. This is not allowed
+     * because certain initializations have not been done. This test is
+     * related to CONFIGURATION-641.
+     */
+    @Test
+    public void testReadCalledDirectly() throws IOException, ConfigurationException
+    {
+        conf = new XMLConfiguration();
+        String content = "<configuration><test>1</test></configuration>";
+        ByteArrayInputStream bis = new ByteArrayInputStream(content.getBytes());
+        try
+        {
+            conf.read(bis);
+            fail("No exception thrown!");
+        }
+        catch (ConfigurationException e)
+        {
+            assertThat(e.getMessage(), containsString("FileHandler"));
+        }
+    }
+
     /**
      * Removes the test output file if it exists.
      */
